hcl2template/types.packer_config.go

```diff
@@ -311,13 +352,92 @@ func (cfg *PackerConfig) evaluateDatasources(skipExecution bool) hcl.Diagnostics
 			})
 			continue
 		}
+
 		ds.value = realValue
 		cfg.Datasources[ref] = ds
 	}
 
+	// Now that most of our data sources have been started and executed, we can
+	// try to execute the ones that depend on other data sources.
+	for ref := range dependencies {
+		_, moreDiags, _ := cfg.recursivelyEvaluateDatasources(ref, dependencies, skipExecution, 0)
+		// Deduplicate diagnostics to prevent recursion messes.
+		cleanedDiags := map[string]*hcl.Diagnostic{}
+		for _, diag := range moreDiags {
+			cleanedDiags[diag.Summary] = diag
+		}
+
+		for _, diag := range cleanedDiags {
+			diags = append(diags, diag)
+		}
+	}
+
 	return diags
 }
 
+func (cfg *PackerConfig) recursivelyEvaluateDatasources(ref DatasourceRef, dependencies map[DatasourceRef][]DatasourceRef, skipExecution bool, depth int) (map[DatasourceRef][]DatasourceRef, hcl.Diagnostics, bool) {
+	var diags hcl.Diagnostics
+	var moreDiags hcl.Diagnostics
+	shouldContinue := true
+
+	if depth > 10 {
+		// Add a comment about recursion.
+		diags = append(diags, &hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Max datasource recursion depth exceeded.",
+			Detail: "An error occured while recursively evaluating data " +
+				"sources. Either your data source depends on more than ten " +
+				"other data sources, or your data sources have a cyclic " +
+				"dependency. Please simplify your config to continue. ",
+		})
+		return dependencies, diags, false
+	}
+
+	ds := cfg.Datasources[ref]
```

Comment:
I think that if we have a `ds.value` here, we could return, no ?

Reply:
(this way datasources are executed only once)

Reply:
Hmm, I'm not seeing this. In my tests, it looks like we don't get to this point at all unless the data sources still need to be executed.

Reply:
Cool !